### PR TITLE
Material inverse optimization using modular NEML2

### PR DESCRIPTION
**Pull request by dewenyushu**
Refs #26920

### CODE REVIEW

**Review by hugary1995 — CHANGES_REQUESTED** [review]
~My only major comment is regarding setting parameters in the `ExecuteNEML2Model` UO. I feel there are two separate features:~
1. ~Collect MOOSE data to use as NEML2 model parameters.~
2. ~Request parameter gradient for certain NEML2 model parameters.~

~Both features are somewhat tied together in your current implementation: When we set a parameter, we immediately set its `requires_grad` to true. We should probably allow the use case where a MOOSE property/variable is used as NEML2 parameter _without_ requesting the corresponding parameter gradient.~

nevermind...

**Comment by hugary1995** [comment]
> My only major comment is regarding setting parameters in the `ExecuteNEML2Model` UO. I feel there are two separate features:
> 
> 1. Collect MOOSE data to use as NEML2 model parameters.
> 2. Request parameter gradient for certain NEML2 model parameters.
> 
> Both features are somewhat tied together in your current implementation: When we set a parameter, we immediately set its `requires_grad` to true. We should probably allow the use case where a MOOSE property/variable is used as NEML2 parameter _without_ requesting the corresponding parameter gradient.

ugh nevermind me... I just realized that indeed you treat these two things separately.

File: modules/solid_mechanics/include/neml2/materials/NEML2ParameterDerivativeToMOOSEMaterialProperty.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  virtual void initQpStatefulProperties() override;
+
+  /// User object managing the execution of the NEML2 model
+  const ExecuteNEML2Model & _execute_neml2_model;
+
+  /// Emitted material property
+  MaterialProperty<T> & _prop;
+
+  /// Initial condition
+  const MaterialProperty<T> * _prop0;
```
[the remark this answers is not in the document]
Comment (dewenyushu) [reply modules/solid_mechanics/include/neml2/materials/NEML2ParameterDerivativeToMOOSEMaterialProperty.h:50]: Yeah I agree. Let me get this removed.

File: modules/optimization/include/vectorpostprocessors/AdjointStrainStressGradInnerProduct.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#pragma once
+
+#include "BatchPropertyDerivative.h"
+#include "ElementOptimizationFunctionInnerProduct.h"
+#include "SymmetricRankTwoTensor.h"
+
+/**
+ * This is a "glue" material that retrieves the batched stress derivative output vector
+ * and uses the output variables to perform the objective stress integration.
+ */
```
[the remark this answers is not in the document]
Comment (dewenyushu) [reply modules/optimization/include/vectorpostprocessors/AdjointStrainStressGradInnerProduct.h:19]: Good catch!

File: modules/combined/test/tests/optimization/invOpt_elasticity/modular/forward_modular.i
[the remark this answers is not in the document]
Comment (dewenyushu) [reply modules/combined/test/tests/optimization/invOpt_elasticity/modular/forward_modular.i:1]: Yeah I'll remove the legacy test case in my next commit

File: modules/combined/test/tests/optimization/invOpt_elasticity/modular/elasticity.i
[the remark this answers is not in the document]
Comment (dewenyushu) [reply modules/combined/test/tests/optimization/invOpt_elasticity/modular/elasticity.i:1]: Removed!

File: modules/solid_mechanics/include/neml2/userobjects/ExecuteNEML2Model.h [hunk 3]
```diff
@@ -66,6 +73,12 @@ class ExecuteNEML2Model : public NEML2ModelInterface<ElementUserObject>
   /// Prevent output and derivative retrieval after construction
   virtual void checkExecutionStage() const final;
 
+  /// Set parameters from parameter UO and/or enable AD
+  virtual void preCompute();
+
+  /// Obtain derivative of output with respect to parameters
+  virtual void postCompute();
```
[the remark this answers is not in the document]
Comment (dewenyushu) [reply modules/solid_mechanics/include/neml2/userobjects/ExecuteNEML2Model.h]: Functions have been renamed

**Comment by dewenyushu** [comment]
Hi @loganharbour , would you mind helping me investigate the issues with the documentation failing in this PR? - I'm wondering if it might be solely related to the code coverage being lower than the required threshold or if there might be other issues? Thanks!

**Comment by loganharbour** [comment]
It's not related to the code coverage. The primary failure can be found here https://civet.inl.gov/job/2404936/, where the app failures are stemming from the moose documentation failure

**Comment by hugary1995** [comment]
It was quite difficult to sift -- it seems like there are 9 tests missing related issues.

**Comment by dewenyushu** [comment]
> It's not related to the code coverage. The primary failure can be found here https://civet.inl.gov/job/2404936/, where the app failures are stemming from the moose documentation failure

Glad to know this is at least not related to code coverage!

**Comment by dewenyushu** [comment]
> It was quite difficult to sift -- it seems like there are 9 tests missing related issues.

Oh that's a good point. However, I only added/changed 7 test cases in this PR. How come there are 9 missing?

**Comment by dewenyushu** [comment]
What is going on with the json file schema differences here (https://civet.inl.gov/job/2418162/)? - I see those failing on my local machine using the latest moose/devel 

Maybe @loganharbour ?

**Comment by hugary1995** [comment]
The fix is on the way: https://github.com/idaholab/moose/pull/28490

**Review by hugary1995 — APPROVED** [review]
New changes since last review also look good to me.

File: modules/solid_mechanics/src/neml2/materials/NEML2ParameterDerivativeToMOOSEMaterialProperty.C [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      "This object requests the computation of the derivative of a NEML2 output variable with "
+      "respect to a NEML2 model parameter and saves the derivative as a MOOSE material property");
+
+  params.addRequiredParam<UserObjectName>("execute_neml2_model_uo",
+                                          "User object managing the execution of the NEML2 model.");
+  params.addRequiredParam<MaterialPropertyName>("moose_material_property",
+                                                "MOOSE material property to emit");
+  params.addRequiredParam<std::string>("neml2_variable", "NEML2 output variable to read");
+  params.addRequiredParam<std::string>(
+      "neml2_parameter_derivative", "Return the derivative of neml2_variable with respect to this");
```
Comment (dschwen) [inline modules/solid_mechanics/src/neml2/materials/NEML2ParameterDerivativeToMOOSEMaterialProperty.C:57]: This should be `neml2_parameter`. with doc string "Parameter to take the derivative with respect to."

Comment (dschwen) [reply modules/solid_mechanics/src/neml2/materials/NEML2ParameterDerivativeToMOOSEMaterialProperty.C:57]: I take that back!

File: modules/solid_mechanics/src/neml2/materials/NEML2ParameterDerivativeToMOOSEMaterialProperty.C [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+  // provide an optional initialization of the moose property (because we don't really know if it is
+  // going to become stateful or not)
+  params.addParam<MaterialPropertyName>("moose_material_property_init",
+                                        "Optional material property as the initial condition");
```
Comment (hugary1995) [inline modules/solid_mechanics/src/neml2/materials/NEML2ParameterDerivativeToMOOSEMaterialProperty.C:62]: ...and remove this.

**Review by dschwen — APPROVED** [review]
This looks fine. I might go back and factor out a common base class for parameter and variable user objects (then both could go into `gather_uos`). Also I need to do some profiling if the `std::vector<torch::Tensor>` approach is ok.

File: modules/solid_mechanics/src/neml2/materials/NEML2ParameterDerivativeToMOOSEMaterialProperty.C (hunk 5 above)
Comment (dewenyushu) [reply modules/solid_mechanics/src/neml2/materials/NEML2ParameterDerivativeToMOOSEMaterialProperty.C:62]: Oh yes nice catch!

**Comment by dewenyushu** [comment]
> This looks fine. I might go back and factor out a common base class for parameter and variable user objects (then both could go into `gather_uos`). Also I need to do some profiling if the `std::vector<torch::Tensor>` approach is ok.

That would be awesome. Thank you!
